File: leetcode/lru-cache.cpp
#include "unordered_map"
using namespace std;

struct Node
{
    int key;
    int value;
    Node *prev, *next;
};
class LRUCache
{
private:
    int capacity;
    unordered_map<int, Node> keyMap;
    Node head, tail;
    void remove(int key)
    {
        keyMap[key].next->prev = keyMap[key].prev;
        keyMap[key].prev->next = keyMap[key].next;
        keyMap.erase(key);
    }
    void appendHead(int key, int value)
    {
        keyMap[key].key = key, keyMap[key].value = value;
        keyMap[key].next = head.next;
        head.next->prev = &keyMap[key];
        head.next = &keyMap[key];
        keyMap[key].prev = &head;
    }

public:
    LRUCache(int capacity) : capacity(capacity)
    {
        head.next = &tail;
        tail.prev = &head;
    }
    int get(int key)
    {
        if (keyMap.find(key) == keyMap.end())
            return -1;
        int val = keyMap[key].value;
        remove(key);
        appendHead(key, val);
        return val;
    }

    void put(int key, int value)
    {
        if (keyMap.find(key) != keyMap.end())
            remove(key);
        else if (keyMap.size() >= capacity)
            remove(tail.prev->key);
        appendHead(key, value);
    }
};

File: leetcode/lru-cache.js
class Node {
    constructor(key, val, parent) {
        this.val = val;
        this.key = key;
        parent && this.attchParent(parent)
    }
    setChild(child) {
        this.next = child;
        child.prev = this;
    }
    attchParent(parent) {
        parent.next = this;
        this.prev = parent;
    }
}
class LRUCache {
    constructor(capacity) {
        this.capacity = capacity;
        this.headNode = new Node('', 0);
        this.tailNode = new Node('', 0, this.headNode);
        this.map = new Map();
    }
    delete(key) {
        let node = this.map.get(key);
        if (!node) return;
        node.prev.setChild(node.next);
        this.map.delete(key);
    }
    attachHead(key, value) {
        const node = new Node(key, value);
        node.setChild(this.headNode.next);
        this.headNode.setChild(node);
        this.map.set(key, node);
    }
    get(key) {
        const node = this.map.get(key);
        if (!node) return -1;
        this.delete(key);
        this.attachHead(key, node.val);
        return this.headNode.next.val;
    }
    put(key, value) {
        if (this.map.get(key)||this.map.size >= this.capacity) {
            this.delete(this.tailNode.prev.key);
        }
        this.attachHead(key, value);
    }
}